// inky/lib/inky.js
var extend = require('util')._extend;
var format = require('util').format;
var values = require('object-values');
var cheerio = require('cheerio');
var getAttrs = require('./util/getAttrs');

module.exports = Inky;

/**
 * Creates a new instance of the Inky parser.
 * @class
 */

function Inky(options) {
  options = options || {};

  // HTML tags for custom components
  this.components = extend({

    button: function (element) {
      var classes = ['button'];
      if (element.attr('class')) {
        classes = classes.concat(element.attr('class').split(' '));
      }

      return format('<table class="%s"><tr><th><a href="#">%s</a></th></tr></table>', classes.join(' '), element.html());
    },

    container: function (element) {
      var classes = ['container'];
      if (element.attr('class')) {
        classes = classes.concat(element.attr('class').split(' '));
      }

      return format('<tr><th></th><th class="%s">%s</th><th></th></tr>', classes.join(' '), element.html());
    },

    column: function (element) {
      var classes = ['column'];
      if (element.attr('class')) {
        classes = classes.concat(element.attr('class').split(' '));
      }

      return format('<div %s class="%s"><table width="100%"><tr><td class="inner"><table width="100%" class="contents">%s</table></td></tr></table></div>', getAttrs(element), classes.join(' '), element.html());
    },

    cover: function (element) {
      return format('<tr><th class="cover">%s</th></tr>', element.html());
    },

    row: function (element) {
      var classes = [];
      if (element.attr('class')) {
        classes = classes.concat(element.attr('class').split(' '));
      }

      return format('<tr><td %s class="%s">%s</td></tr>', getAttrs(element), classes.join(' '), element.html());
    },

    wrapper: function (element) {
      return format('<table class="wrapper">%s</table>', element.html());
    }

  }, options.components || {});

}

/**
 * Awww yiss. Kickstarts the whole parser. Takes in HTML loaded via Cheerio as an argument, checks if there are any custom components. If there are, it replaces the nested components, traverses the DOM and replaces them with email markup.
 * @param {object} $ - Input HTML as a Cheerio object
 * @returns {object} Modified HTML as a Cheerio object
 */
Inky.prototype.releaseTheKraken = function ($) {

  // This large compound selector looks for any custom tag loaded into Inky
  // <center> is an exception: the selector is center:not([data-parsed])
  // Otherwise the parser gets caught in an infinite loop where it continually tries to process the same <center> tags
  var tags = Object.keys(this.components).map(function (tag) {
    return tag;
  }).join(', ');

  // Because the structure of the DOM constantly shifts, we carefully go through each custom tag one at a time, until there are no more custom tags to parse
  while ($(tags).length > 0) {
    var elem = $(tags).eq(0);
    var newHtml = this.components[$(tags)[0].name].call(this, elem);
    elem.replaceWith(newHtml);
  }

  return $;

}
